x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1148,32 +1149,37 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
     static class SkipQueryIfFoldingProjection extends OptimizerRule<LogicalPlan> {
         @Override
         protected LogicalPlan rule(LogicalPlan plan) {
-            Holder<LocalRelation> optimizedPlan = new Holder<>();
-            plan.forEachDown(Project.class, p -> {
-                List<Object> values = extractConstants(p.projections());
-                if (values.size() == p.projections().size() && (p.child() instanceof EsRelation) == false &&
-                    isNotQueryWithFromClauseAndFilterFoldedToFalse(p)) {
-                    optimizedPlan.set(new LocalRelation(p.source(), new SingletonExecutable(p.output(), values.toArray())));
-                }
-            });
+            LogicalPlan relation = plan
+                .collectFirstDown(p -> p.children().isEmpty() ? Optional.of(p) : Optional.empty())
+                .get();
 
-            if (optimizedPlan.get() != null) {
-                return optimizedPlan.get();
-            }
+            // exclude LocalRelations that have been introduced by earlier optimizations (skipped ESRelations)
+            boolean isNonSkippedLocalRelation = relation instanceof LocalRelation &&
+                ((LocalRelation) relation).executable() instanceof EmptyExecutable == false;
+
+            Optional<LogicalPlan> optimized = plan.collectFirstDown(p -> {
+                Optional<List<Object>> foldedValues = Optional.empty();
+
+                if (p instanceof Project && isNonSkippedLocalRelation) {
+                    foldedValues = Optional.of(extractConstants(((Project) p).projections()));
+                } else if (p instanceof Aggregate) {
+                    Aggregate a = (Aggregate) p;
+                    List<Object> folded = extractConstants(a.aggregates());
+
+                    boolean onlyConstantAggregations = relation instanceof EsRelation &&
```

Comment:
We should try to use a consistent code style and one of the practices is the use of `&&` and `||` at the start of the line, not at the end, in case of multiline conditions.